### PR TITLE
Progress bars for extendedP2proc()

### PR DESCRIPTION
**Pull request by rrydbirk**
Due to experiencing stalling during extendedP2proc, I've added progress bars to aid in visual assessment of progress. Specifically, since processing of valid pathways can take a considerable amount of time, I added a progress bar from pbapply package. Additionally, I added additional progress messages for the different processes in extendedP2proc.

### CODE REVIEW

File: R/Pagoda2.r [hunk 1]
```diff
@@ -1718,11 +1718,44 @@ Pagoda2 <- setRefClass(
       return(invisible(list(d=d,m=m,gpcs=gpcs)))
     },
 
+#test consistency
+tPO=function(setenv=go.env, verbose=T, n.cores=1) {
```
Comment (VPetukhov) [inline R/Pagoda2.r]: This function is never used in the code. What should it do? And please, use meaningful function names.

File: R/Pagoda2.r [hunk 2]
```diff
@@ -1718,11 +1718,44 @@ Pagoda2 <- setRefClass(
...
+      invisible(require("pbmcapply"))
```
Comment (VPetukhov) [inline R/Pagoda2.r]: `require` must never be used inside a package. Instead, required package should be added to DESCRIPTION and function call should be performed through full names: `pbmcapply::pbmclapply`. Btw, I'd strongly suggest to use `pbapply` package instead.

File: R/Pagoda2.r [hunk 3]
```diff
@@ -1718,11 +1718,44 @@ Pagoda2 <- setRefClass(
...
+        if(verbose) cat(" done\n")
+        
+          return(invisible(length(gsl)))         
+  },
 
     # test pathway overdispersion
     # this is a compressed version of the PAGODA1 approach
     # env - pathway to gene environment
     testPathwayOverdispersion=function(setenv, type='counts', max.pathway.size=1e3, min.pathway.size=10, n.randomizations=5, verbose=FALSE, score.alpha=0.05, plot=FALSE, cells=NULL,adjusted.pvalues=TRUE,z.score = qnorm(0.05/2, lower.tail = FALSE), use.oe.scale = FALSE, return.table=FALSE,name='pathwayPCA',correlation.distance.threshold=0.2,loading.distance.threshold=0.01,top.aspects=Inf,recalculate.pca=FALSE,save.pca=TRUE) {
+      require("pbmcapply")
```
Comment (VPetukhov) [inline R/Pagoda2.r]: Again, `require`

File: R/Pagoda2.r [hunk 4]
```diff
@@ -1784,19 +1821,20 @@ Pagoda2 <- setRefClass(
             rownames(pcs$rotation) <- colnames(x)[lab];
           } # don't bother otherwise - it's not significant
           return(list(xp=pcs,z=z,n=ngenes))
-        }, n.cores = n.cores,mc.preschedule=T)
-        if(save.pca) {
+        }, mc.cores = n.cores,mc.preschedule=T)
+                                
+if(save.pca) {
```
Comment (VPetukhov) [inline R/Pagoda2.r]: Broken formatting

File: R/pipelineHelpers.R [hunk 5]
```diff
@@ -202,16 +202,20 @@ p2.generate.dr.go <- function(r) {
   }
   
   # translate gene names to ids
+  message("Preparing GO environment ", appendLF=F)
```
Comment (VPetukhov) [inline R/pipelineHelpers.R]: All diagnostic messages should be displayed only with `verbose==TRUE`

File: R/Pagoda2.r [hunk 6]
```diff
@@ -1741,22 +1774,26 @@ Pagoda2 <- setRefClass(
...
 
         if(verbose) {
-          message("processing ", length(gsl), " valid pathways")
+          message("Processing ", length(gsl), " valid pathways")
         }
 
         cm <- Matrix::colMeans(x)
 
-        pwpca <- papply(gsl, function(sn) {
+pwpca <- pbmclapply(gsl, function(sn) {
```
Comment (VPetukhov) [inline R/Pagoda2.r]: There must be a way to disable all logs. I suggest to use check `lapply.func <- if (verbose) pbmclapply else plapply`

**Comment by VPetukhov** [comment]
Thank you for contribution, Rasmus! There are some style issues, which must be addressed before merging the request though. Please, see the comments.

File: R/Pagoda2.r (hunk 1 above)
Comment (rrydbirk) [reply R/Pagoda2.r]: Sorry, this is my first pull, this was meant for my eyes only. I didn't realise you would see additional changes made after the pull request. I will remove this.

File: R/Pagoda2.r (hunk 6 above)
Comment (rrydbirk) [reply R/Pagoda2.r]: I'm sorry, Victor, I don't understand your point. How is this chunk related to logs?
Also, from what I understand, pbapply will only do parallel if cl > 1, so then check is not necessary, right?

File: R/Pagoda2.r (hunk 2 above)
Comment (rrydbirk) [reply R/Pagoda2.r]: Thanks for pointing this out, I was unaware of proper standards using DESCRIPTION. Also, using pbapply instead.

File: R/Pagoda2.r (hunk 1 above)
Comment (rrydbirk) [reply R/Pagoda2.r]: This was meant for my eyes only, pushed to wrong branch, sorry.

File: R/Pagoda2.r (hunk 6 above)
Comment (rrydbirk) [reply R/Pagoda2.r]: I've created pwpca.fun and then check for verbose as suggested. Will this do?

**Comment by evanbiederstedt** [comment]
Hi @rrydbirk 

Do you happen to know whether this PR is still relevant? 

If so, it might be useful for me if you do another PR against the latest `dev` branch. 

Thanks, Evan

**Comment by rrydbirk** [comment]
> Hi @rrydbirk
> 
> Do you happen to know whether this PR is still relevant?
> 
> If so, it might be useful for me if you do another PR against the latest `dev` branch.
> 
> Thanks, Evan

Hi Evan

No, I created the PR since my old cluster had problems with extendedP2proc. This is not relevant anymore.

/Rasmus

**Comment by evanbiederstedt** [comment]
Hi @rrydbirk 

Thanks for looking into this. Also, I think this issue is relevant: https://github.com/kharchenkolab/pagoda2/issues/83
Can it be closed? 

Thanks, Evan

**Pull request closed without merging**